doc/optimizers.md

```diff
@@ -1563,6 +1563,68 @@ optimizer.Optimize(f, coordinates);
  * [SGD in Wikipedia](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)
  * [Differentiable separable functions](#differentiable-separable-functions)
 
+## NSGA2
+
+*An optimizer for arbitrary multi-objective functions.*
+
+NSGA2 (Non-dominated Sorting Genetic Algorithm - II) is a multi-objective
+optimization algorithm. The algorithm works by generating a candidate population
+from a fixed starting point. At each stage of optimization, a new population of
+children is generated. This new population along with its predecessor is sorted
+using non-domination as the metric. Following this, the population is further
+segregated into fronts. A new population is generated from these fronts having
+size equal to that of the starting population.
+
+#### Constructors
+
+ * `NSGA2()`
+ * `NSGA2(`_`populationSize, maxGenerations, crossoverProb, mutationProb, mutationStrength, epsilon, lowerBound, upperBound`_`)`
+
+#### Attributes
+
+| **type** | **name** | **description** | **default** |
+|----------|----------|-----------------|-------------|
+| `size_t` | **`populationSize`** | The number of candidates in the population. This should be at least 4 in size and a multiple of 4. | `100` |
+| `size_t` | **`maxGenerations`** | The maximum number of generations allowed for NSGA2. | `2000` |
+| `double` | **`crossoverProb`** | Probability that a crossover will occur. | `0.6` |
+| `double` | **`mutationProb`** | Probability that a weight will get mutated. | `0.3` |
+| `double` | **`mutationStrength`** | The range of mutation noise to be added. This range is between 0 and mutationStrength. | `0.001` |
+| `double` | **`epsilon`** | The final value of the objective function for termination. If set to negative value, tolerance is not considered. | `1e-6` |
+| `arma::vec` | **`lowerBound`** | Lower bound of the coordinates of the initial population. | `1` |
+| `arma::vec` | **`upperBound`** | Lower bound of the coordinates of the initial population. | `1` |
+
+
+Attributes of the optimizer may also be changed via the member methods
+`PopulationSize()`, `MaxGenerations()`, `CrossoverRate()`, `MutationProbability()`, `MutationStrength()`, `Epsilon()`, `LowerBound()` and `UpperBound()`.
+
+#### Examples:
+
+<details open>
+<summary>Click to collapse/expand example code.
+</summary>
+
+```c++
+SchafferFunctionN1<arma::mat> SCH;
+arma::vec lowerBound("-1000 -1000");
+arma::vec upperBound("1000 1000");
+NSGA2 opt(20, 5000, 0.5, 0.5, 1e-3, 1e-6, lowerBound, upperBound);
+
+typedef decltype(SCH.objectiveA) ObjectiveTypeA;
+typedef decltype(SCH.objectiveB) ObjectiveTypeB;
+
+arma::mat coords = SCH.GetInitialPoint();
+std::tuple<ObjectiveTypeA, ObjectiveTypeB> objectives = SCH.GetObjectives();
+
+std::vector<arma::mat> bestFront = opt.Optimize(objectives, coords);
```

Comment:
It looks like `Optimize()` here differs from other ensmallen optimizers in that it returns the best front, not the objective.  Do you think we should point that out in the multi-objective optimizers section?

Reply:
Sure, I shall add it.